File: leetcode/202501/ReverseInteger/ReverseInteger.rs
// @Introduce  : 
// @File       : ReverseInteger.rs
// @Author     : ryrl
// @Email      : ryrl970311@gmail.com
// @Time       : 2025/01/10 17:20
// @Description:

fn main() {
    let x = 123;
    let res = reverse(x);
    println!("{}", res);
}

pub fn reverse(x: i32) -> i32 {
    let mut res: i32 = 0;
    let mut cur: i32 = x;
        
    while cur != 0 {
        match res.checked_mul(10) {
            None => return 0,
            Some(tmp) => match tmp.checked_add(cur % 10) {
                None => return 0,
                Some(fine) => {
                    res = fine;
                }
            } 
        }
        cur = cur / 10;
    }
    res
}

File: leetcode/Daily/ReverseInteger/ReverseInteger.go
// @Introduce  :
// @File       : ReverseInteger.go
// @Author     : ryrl
// @Email      : ryrl970311@gmail.com
// @Time       : 2025/01/10 14:45
// @Description:

package main

import "fmt"

func main() {
	var x int = -123
	fmt.Printf("-123 %% 10: %v\n", x % 10)
}

func reverse(x int) int {
	var res int
	for x != 0 {
		res = res * 10 + x % 10
		x /= 10
	}
	if res > 1<<31-1 || res < -1<<31 {
		return 0
	}
	return res
}